zebra-chain/src/transaction/serialize.rs

```diff
@@ -544,10 +544,33 @@ impl ZcashDeserialize for Transaction {
             (header & LOW_31_BITS, header >> 31 != 0)
         };
 
-        // Consensus rules:
+        // # Consensus
+        //
+        // The next rules apply for different transaction versions as follows:
+        //
+        // [Pre-Overwinter]: Transactions version 1 and 2.
+        // [Overwinter onward]: Transactions version 3 and above.
+        // [Overwinter only, pre-Sapling]: Transactions version 3.
+        // [Sapling to Canopy inclusive, pre-NU5]: Transactions version 4.
+        // [NU5 onward]: Transactions version 4 and above.
+        //
         // > The transaction version number MUST be greater than or equal to 1.
-        // >
-        // > The overwintered flag MUST NOT be set for version 1 and 2 transactions.
+        //
+        // > [Pre-Overwinter] The fOverwintered fag MUST NOT be set.
+        //
+        // > [Overwinter onward] The version group ID MUST be recognized.
+        //
+        // > [Overwinter onward] The fOverwintered flag MUST be set.
+        //
+        // > [Overwinter only, pre-Sapling] The transaction version number MUST be 3,
+        // > and the version group ID MUST be 0x03C48270.
+        //
+        // > [Sapling to Canopy inclusive, pre-NU5] The transaction version number MUST be 4,
+        // > and the version group ID MUST be 0x892F2085.
+        //
+        // > [NU5 onward] The transaction version number MUST be 4 or 5.
```

Comment:
I think this is not checked here (it parses whatever version the transaction is), neither where the old comment was. We kind of test it in a roundabout way.

Here we only allow version 4 or 5 (other versions are checkpointed and never fully verified). This already satisfies the consensus rule.

https://github.com/ZcashFoundation/zebra/blob/154fc0a2fb01a5d9d45f0cb5d3ebc25b5624dc4b/zebra-consensus/src/transaction.rs#L360

The place where the old comment was checks that, for example, if transaction is 4, then it must be Sapling-onward. Which is not exactly the same thing.

What do you think of moving this rule to the line above?

Reply:
Hmm, I am not so sure about it. It is true that here we are not checking if the transaction is 4 or 5 but when we get a transaction 4 or 5 we make sure the group id is correct which is the second part of those 2 rules.

Reply:
What do you think about something like https://github.com/ZcashFoundation/zebra/pull/3456/commits/124760e4ffe821caee60830e79cfbd8231690a30 ?

Reply:
Looks good!

Reply:
Checkpointing allows us to ignore most rules for pre-Canopy network upgrades.

But we still need to check that Canopy and NU5 have the correct transaction versions. That is done here:

https://github.com/ZcashFoundation/zebra/blob/fa071562fd0b891f29143e4cc8d17b34b4b1cf10/zebra-consensus/src/transaction.rs#L547-L550

https://github.com/ZcashFoundation/zebra/blob/fa071562fd0b891f29143e4cc8d17b34b4b1cf10/zebra-consensus/src/transaction.rs#L636-L637

Reply:
In particular, Canopy can't have version 5 transactions, so we still need to check these rules somewhere. (Parsing and sighash are not enough.)

Reply:
So i think we should keep those docs. I re added them updating the format at https://github.com/ZcashFoundation/zebra/pull/3456/commits/eb20e78c5fd9c01055a91bcb3d50990ac07dbc6a